File: LeetCode/1849. Splitting a String Into Descending Consecutive Values/solution.cpp
using LL = long long;

class Solution {
public:
    bool dfs(string &s, LL pre, int i){
        if (i == s.size())
            return true;
        LL cur = 0;
        while (i < s.size()){
            cur *= 10;
            cur += s[i++]-'0';
            if (cur == pre-1 && dfs(s, cur, i))
                return true;
            if (cur > pre-1)
                return false;
        }
        return false;
    }

    bool splitString(string s) {
        LL cur = 0;
        for (int i = 0; i+1 < s.size();){
            cur *= 10;
            cur += s[i++]-'0';
            if (cur > (LL) INT_MAX*10)
                break;
            if (dfs(s, cur, i))
                return true;
        }
        return false;
    }
};

File: LeetCode/1849. Splitting a String Into Descending Consecutive Values/solution.py
from functools import lru_cache
class Solution:
    def splitString(self, s: str) -> bool:
        n = len(s)
        @lru_cache(None)
        def dfs(pre, i):
            if i == n:
                return True
            for j in range(i+1, n+1):
                if int(s[i:j]) == pre-1 and dfs(pre-1, j):
                    return True
            return False
        return any(dfs(int(s[:i]), i) for i in range(1,n))